### PR TITLE
Skip entire subtrees while scanning for source files

### LINKED ISSUE #6240: Excluded files affects scanning speed
## Expected Behavior
Detekt runs fast with skipping excluded folders and files.

## Observed Behavior
Detekt checks all files, even if they are excluded and it takes a very long time

## Steps to Reproduce
-

## Context
I have a big android project with many modules. Each module folder contains **build** folder generated by gradle, which contains huge amount of generated files. All build directories are added to excludes with `**/build/**`.
If project is fresh without **build** directories, detekt task ends in 30 seconds.
When **build**  folders are present, detekt task ends in 5-10 minutes depends on machine.
I looked at the code of detekt and found out the problem is in include-exclude mechanism itself. Detekt takes every file even in and checks it path to include\exclude paths.
So if project contains giant amount of files in excluded directories, it significantly increases scanning time.

## Your Environment
* Version of detekt used:  v1.22.0, v1.23.0
* Version of Gradle used (if applicable): 8.1.1
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: Macos 13,  Windows 10
* Link to your project (if it's a public repository): N/A

### PR DESCRIPTION
**Pull request by marschwar**
fixes #6240

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt [hunk 1]
```diff
@@ -21,6 +26,10 @@ class KtTreeCompiler(
     private val basePath: Path? = projectSpec.basePath
     private val pathFilters: PathFilters? =
         PathFilters.of(projectSpec.includes.toList(), projectSpec.excludes.toList())
+    private val subtreeExcludes = projectSpec.excludes
+        .filter { it.endsWith("/**") }
+        .map { pathMatcher(it) }
+        .toList()
```
Comment (marschwar) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:32]: I am not sure that this is the ideal way to check for subtree globs

Comment (illuzor) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:32]: Exceptions don't have to always end with /**
You can specify specific folders. For example `app/build`

Comment (marschwar) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:32]: Thank you. Depending on the outcome of the conflict with the other work that is being done at the moment (see above), I will update this.

**Comment by 3flex** [comment]
I've replaced https://github.com/detekt/detekt/pull/6271 with https://github.com/detekt/detekt/pull/6310 which keeps `KtTreeCompiler`, so this won't be blocked after all.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt [hunk 2]
```diff
@@ -57,12 +58,43 @@ class KtTreeCompiler(
...
+                settings.debug { "Ignoring file '$file'" }
+            } else {
+                _collected.add(file)
+            }
+            return FileVisitResult.CONTINUE
+        }
+
+        override fun visitFileFailed(file: Path, exc: IOException): FileVisitResult {
+            settings.error("Error visiting file '$file'.", exc)
+            return FileVisitResult.TERMINATE
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:88]: Should we just throw the exception? If we have an exception at this point the issue is really big and if we just ignore the scan will be only partial.

I know that this is a common pattern on detekt: Try to no fail. But this make the work of debugging issues way harder.

**Review by BraisGabin — APPROVED** [review]
LGTM.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt (hunk 2 above)
Comment (marschwar) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:88]: That is a good point. I have changed this.

**Pull request merged**